### ISSUE TITLE
[Arm64/Unix] Corefx *.dll manifest definition does not match the assembly reference

### ISSUE THREAD

**Issue by sdmaclea** [dotnet/coreclr#11764]
`7a75598` I failing 100% of tests on [Arm64/Unix]
Last known good version `bad6729`

Needs triage

@dotnet/arm64-contrib I will look at this next week.

**Comment by BruceForstall** [comment]
fyi @dotnet/jit-contrib

**Comment by sdmaclea** [comment]
Looks like the first patch which caused significant regression was  01fe990 Which was the Update CoreFx to preview2-25319-01 (#11736) unfortunately.  

However this only caused 80% of the tests to fail, there is apparently another breaking commit.

The failures I observed seemed to be missing dependencies.
```
Could not load file or assembly 'System.Runtime.Extensions`
```

**Comment by BruceForstall** [comment]
It seems like we should get these "Update CoreFx" and similar changes to run arm64 tests as a general rule. cc @gkhanna79 .

**Comment by sdmaclea** [comment]
I looked at this case which was missing System.Console.  It looks like the DLL search path is not including next to `corerun`.  It is likely this means this is only working on other platforms because it is hitting in the nuget package cache.  

```
bin/tests/pr1/Tests/coreoverlay/corerun bin/tests/pr1/baseservices/threading//generics/syncdelegate/GThread26/GThread26.exe

Unhandled Exception: System.IO.FileNotFoundException: Could not load file or assembly 'System.Console, Version=4.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. The system cannot find the file specified.
```

```
strace -e trace=openat ".../bin/tests/pr1/Tests/coreoverlay/corerun bin/tests/pr1/baseservices/threading//generics/syncdelegate/GThread26/GThread26.exe 2>&1 | grep vmjenkins
openat(AT_FDCWD, ".../bin/tests/pr1/Tests/coreoverlay", O_RDONLY|O_NONBLOCK|O_DIRECTORY|O_CLOEXEC) = 3
openat(AT_FDCWD, ".../bin/tests/pr1/Tests/coreoverlay/libcoreclr.so", O_RDONLY|O_CLOEXEC) = 3
openat(AT_FDCWD, ".../bin/tests/pr1/Tests/coreoverlay/System.Private.CoreLib.dll", O_RDONLY) = 9
openat(AT_FDCWD, ".../bin/tests/pr1/baseservices/threading/generics/syncdelegate/GThread26/System.Native", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, ".../bin/tests/pr1/Tests/coreoverlay/System.Native", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, ".../bin/tests/pr1/Tests/coreoverlay/System.Native", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, ".../bin/tests/pr1/Tests/coreoverlay/libSystem.Native.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, ".../bin/tests/pr1/baseservices/threading/generics/syncdelegate/GThread26/libSystem.Native.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, ".../bin/tests/pr1/Tests/coreoverlay/libSystem.Native.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, ".../bin/tests/pr1/Tests/coreoverlay/System.Native.so", O_RDONLY|O_CLOEXEC) = 10
openat(AT_FDCWD, ".../bin/tests/pr1/baseservices/threading/generics/syncdelegate/GThread26/libcoreclr", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, ".../bin/tests/pr1/Tests/coreoverlay/libcoreclr", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, ".../bin/tests/pr1/Tests/coreoverlay/libcoreclr", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, ".../bin/tests/pr1/Tests/coreoverlay/liblibcoreclr.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, ".../bin/tests/pr1/baseservices/threading/generics/syncdelegate/GThread26/liblibcoreclr.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, ".../bin/tests/pr1/Tests/coreoverlay/liblibcoreclr.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, ".../bin/tests/pr1/baseservices/threading/generics/syncdelegate/GThread26/GThread26.exe", O_RDONLY) = 10
openat(AT_FDCWD, ".../bin/tests/pr1/Tests/coreoverlay/System.Runtime.ni.dll", O_RDONLY) = 11
openat(AT_FDCWD, ".../bin/tests/pr1/Tests/coreoverlay/System.Runtime.ni.dll", O_RDONLY) = 13
openat(AT_FDCWD, ".../bin/tests/pr1/Tests/coreoverlay/libclrjit.so", O_RDONLY|O_CLOEXEC) = 14
openat(AT_FDCWD, ".../bin/tests/pr1/baseservices/threading/generics/syncdelegate/GThread26/System.Console.ni.dll", O_RDONLY) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, ".../bin/tests/pr1/baseservices/threading/generics/syncdelegate/GThread26/System.Console.ni.exe", O_RDONLY) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, ".../bin/tests/pr1/baseservices/threading/generics/syncdelegate/GThread26/System.Console.dll", O_RDONLY) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, ".../bin/tests/pr1/baseservices/threading/generics/syncdelegate/GThread26/System.Console.exe", O_RDONLY) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, ".../bin/tests/pr1/baseservices/threading/generics/syncdelegate/GThread26/System.Globalization.Native", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, ".../bin/tests/pr1/Tests/coreoverlay/System.Globalization.Native", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, ".../bin/tests/pr1/Tests/coreoverlay/System.Globalization.Native", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, ".../bin/tests/pr1/Tests/coreoverlay/libSystem.Globalization.Native.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, ".../bin/tests/pr1/baseservices/threading/generics/syncdelegate/GThread26/libSystem.Globalization.Native.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, ".../bin/tests/pr1/Tests/coreoverlay/libSystem.Globalization.Native.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, ".../bin/tests/pr1/Tests/coreoverlay/System.Globalization.Native.so", O_RDONLY|O_CLOEXEC) = 16
```

**Comment by sdmaclea** [comment]
The other interesting thing is that after bin/tests/pr1/Tests/coreoverlay/libcoreclr.so is sucessfully loaded there is another attempt to load it.  Maybe time is not ordered because this is a threading test.


@jkotas Do you know why the corerun dll search path would not include the corerun directory?

**Comment by jkotas** [comment]
I do not know

**Comment by sdmaclea** [comment]
@jkotas,@janvorli,@gkhanna79 The Trusted Platform Assemblies is not being populated correctly.  

It looks like this may be related to #10525, but it also looks like `GetTpaList()` in corerun and coreconsole are not thread safe and may not be initialized correctly if two threads call.

**Comment by jkotas** [comment]
There should not ever be two thread calling it in corerun or coreconsole. There should be a single thread running where CoreCLR is getting initialized in corerun or console.

**Comment by sdmaclea** [comment]
OK I completely missed the issue.  The new corefx dll s are failing to crossgen.  To complicate matters, the one byte error code is aliasing with expected failures due to native windows dlls which are handled by deleting the platform incompatible dll.

The errors look like the following

```
Error:./System.IO.FileSystem.Watcher.dll.log
Error: Could not load file or assembly 'System.ComponentModel.Primitives, Version=4.2.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040 (FUSION_E_REF_DEF_MISMATCH))
Error compiling ./System.IO.FileSystem.Watcher.dll: Could not find or load a specific file. (Exception from HRESULT: 0x80131621 (COR_E_FILELOAD))
Error: compilation failed for "./System.IO.FileSystem.Watcher.dll" (0x80131621)
Microsoft (R) CoreCLR Native Image Generator - Version 4.5.22220.0
Copyright (c) Microsoft Corporation.  All rights reserved.
```

The following managed assemblies are affected

```
jit-dasm.dll
Microsoft.Diagnostics.Tracing.TraceEvent.dll
Microsoft.DotNet.Cli.Utils.dll
Microsoft.DotNet.ProjectModel.dll
NuGet.Common.dll
NuGet.Protocol.Core.v3.dll
System.ComponentModel.TypeConverter.dll
System.Data.Common.dll
System.Data.Odbc.dll
System.Data.SqlClient.dll
System.Diagnostics.FileVersionInfo.dll
System.Diagnostics.Process.dll
System.Diagnostics.StackTrace.dll
System.Diagnostics.TextWriterTraceListener.dll
System.Diagnostics.TraceSource.dll
System.IO.FileSystem.AccessControl.dll
System.IO.FileSystem.Watcher.dll
System.IO.IsolatedStorage.dll
System.IO.Pipes.dll
System.Memory.dll
System.Net.Http.dll
System.Net.HttpListener.dll
System.Net.Mail.dll
System.Net.NetworkInformation.dll
System.Net.Ping.dll
System.Net.Requests.dll
System.Net.Security.dll
System.Net.WebClient.dll
System.Net.WebSockets.Client.dll
System.Private.DataContractSerialization.dll
System.Private.Xml.dll
System.Private.Xml.Linq.dll
System.Runtime.Extensions.dll
System.Security.AccessControl.dll
System.Security.Cryptography.Algorithms.dll
System.Security.Cryptography.Cng.dll
System.Security.Cryptography.Csp.dll
System.Security.Cryptography.Encoding.dll
System.Security.Cryptography.OpenSsl.dll
System.Security.Cryptography.X509Certificates.dll
System.Security.Permissions.dll
System.Security.Principal.Windows.dll
System.Threading.Thread.dll
xunit.performance.api.dll
```

**Comment by sdmaclea** [comment]
@jkotas @RussKeldorph I just changed the title to describe the real problem.  Is this a corefx issue?  Is it a crossgen issue?  Who should I engage in this?

**Comment by sdmaclea** [comment]
This is also causing the following tests to fails
```
GC/API/GC/Collect_Forced_1/Collect_Forced_1.sh
GC/API/GC/Collect_Forced_2/Collect_Forced_2.sh
GC/API/GC/Collect_Forced_3/Collect_Forced_3.sh
JIT/Performance/CodeQuality/Serialization/Deserialize.Deserialize.sh
JIT/Performance/CodeQuality/Serialization/Serialize/Serialize.sh
JIT/superpmi/superpmicollect/superpmicollect.sh
```

They are only affected by 
```
System.ComponentModel.Primitives
System.Text.RegularExpressions
```

**Comment by sdmaclea** [comment]
@weshaggard Is this the same issue you are trying to fix with corefx in https://github.com/dotnet/corefx/issues/20025

**Comment by weshaggard** [comment]
The error I'm fixing in https://github.com/dotnet/corefx/pull/20025 is about corefx's consumption of dependencies, I don't think that should impact coreclr. However I am fixing a similar issue in the coreclr update PR https://github.com/dotnet/coreclr/pull/11809. I'm not sure if it is the exact issue but it is similar.

**Comment by sdmaclea** [comment]
This was fixed by #11809
